File: CSES/Trees/Subordinates.py
I = lambda: input()
II = lambda: int(input())
MII = lambda: map(int, input().split())
LI = lambda: list(input().split())
LII = lambda: list(map(int, input().split()))
LGMII = lambda: map(lambda x: int(x) - 1, input().split())
LGLII = lambda: list(map(lambda x: int(x) - 1, input().split()))
from collections import defaultdict

inf = float('inf')

def solve():
    n = II()
    edges = LII()
    adj = defaultdict(list[int])
    for i in range(len(edges)):
        adj[edges[i]-1].append(i+1)
    ans = [0]*n
    def go(node):
        for x in adj[node]:
            go(x)
            ans[node]+=ans[x]+1
    go(0)
    print(*ans)
solve()

File: CSES/Trees/Subordinates.cpp

#include<bits/stdc++.h>
using namespace std;

int main(){
    int n;
    cin>>n;
    vector<int> parents(n-1,0);
    for(int i=0; i<n-1; i++){
        cin>>parents[i];
    }
    vector<vector<int>> adj(n);
    int idx = 1;
    for(auto x:parents){
        adj[x-1].push_back(idx);
        idx++;
    }
    vector<int>ans(n,0);
    function<void(int)> go = [&](int node){
        for(auto x:adj[node]){
            go(x);
            ans[node]+=ans[x]+1;
        }
    };
    go(0);
    for(auto x:ans){
        cout<<x<<" ";
    }
}
